### ISSUE TITLE
Different behaviour of `brace_linter()` inside and outside of function call

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#1829]
When outside a function call, `brace_linter()` lints both the opening and the closing braces:

``` r
lint(text = "1:4 %>% { sum(.) }", linters = brace_linter())
#> <text>:1:9: style: [brace_linter] Opening curly braces should never go on their own line and should always be followed by a new line.
#> 1:4 %>% { sum(.) }
#>         ^
#> <text>:1:18: style: [brace_linter] Closing curly-braces should always be on their own line, unless they are followed by an else.
#> 1:4 %>% { sum(.) }
#>                  ^
```

<sup>Created on 2022-12-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

When inside a function call, `brace_linter()` lints only the opening brace for the same expression:

```r
lint(text = "local(1:4 %>% { sum(.) })", linters = brace_linter())
#> <text>:1:15: style: [brace_linter] Opening curly braces should never go on their own line and should always be followed by a new line.
#> local(1:4 %>% { sum(.) })
#>               ^
```

<sup>Created on 2022-12-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Is this expected?

---

Making sure they produce identical results:

``` r
library(lintr)
library(magrittr)

identical(1:4 %>% { sum(.) }, local(1:4 %>% { sum(.) }))
#> [1] TRUE
```

**Comment by AshesITR** [comment]
Yes, we also allow `})` without mentioning it in the message.
 e.g.
```r
test_that("math works", {
  expect_equal(1 + 1, 2)
})
```

**Comment by MichaelChirico** [comment]
I guess our intention was to allow something like:

```r
local({
  1:4 %>% sum()
})
```

In this case, the `}` is not attached to the same expression as the `)`, maybe that's something we could fix?

I also wonder about something like:

```r
local({
  1:4 %>% sum()})
```

Seems like it should lint -- even though `}` and `)` are together, `}` should still go on its own line.
